File: app/src/main/java/com/dujiajun/courseblock/MainActivity.java
package com.dujiajun.courseblock;

import android.content.ComponentName;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.preference.PreferenceManager;

import com.dujiajun.courseblock.helper.CourseManager;
import com.dujiajun.courseblock.helper.WeekManager;
import com.dujiajun.courseblock.model.Course;
import com.zhuangfei.timetable.TimetableView;
import com.zhuangfei.timetable.listener.ISchedule;
import com.zhuangfei.timetable.listener.OnSlideBuildAdapter;
import com.zhuangfei.timetable.listener.OnSpaceItemClickAdapter;
import com.zhuangfei.timetable.view.WeekView;

public class MainActivity extends AppCompatActivity {

    private SharedPreferences preferences;
    private CourseManager courseManager;
    private TimetableView timetableView;
    private WeekView weekView;
    private WeekManager weekManager;
    private int showWeek = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        preferences = PreferenceManager.getDefaultSharedPreferences(this);
        weekManager = WeekManager.getInstance(this);

        timetableView = findViewById(R.id.id_timetableView);
        weekView = findViewById(R.id.id_weekview);
        weekView.curWeek(weekManager.getCurWeek())
                .callback(week -> {
                    int cur = timetableView.curWeek();
                    timetableView.onDateBuildListener()
                            .onUpdateDate(cur, week);
                    timetableView.changeWeekOnly(week);
                    showWeek = week;
                })
                .hideLeftLayout()
                .itemCount(Course.MAX_WEEKS)
                .isShow(false).showView();

        boolean show_weekend = preferences.getBoolean("show_weekend", true);
        boolean show_not_cur_week = preferences.getBoolean("show_not_cur_week", true);
        boolean show_time = preferences.getBoolean("show_course_time", true);

        timetableView.curWeek(weekManager.getCurWeek())
                .isShowNotCurWeek(show_not_cur_week)
                .isShowWeekends(show_weekend)
                .maxSlideItem(Course.MAX_STEPS)
                .callback(new OnSpaceItemClickAdapter() {
                    @Override
                    public void onSpaceItemClick(int day, int start) {
                        Intent intent = new Intent(MainActivity.this, CourseActivity.class);
                        intent.putExtra("day", day + 1);
                        intent.putExtra("start", start);
                        intent.putExtra("week", showWeek);
                        intent.putExtra("action", CourseActivity.ACTION_INSERT);
                        startActivity(intent);
                    }
                })
                .callback((ISchedule.OnItemLongClickListener) (v, day, start) -> {
                    Course c = courseManager.findCourseByDayAndStart(showWeek, day, start);
                    if (c != null) {
                        Intent intent = new Intent(MainActivity.this, CourseActivity.class);
                        intent.putExtra("action", CourseActivity.ACTION_DETAIL);
                        intent.putExtra("course", c.getId());
                        startActivity(intent);
                    }
                });
        if (show_time)
            showTime();
        courseManager = CourseManager.getInstance(getApplicationContext());
        updateView();
        updateWeek();
    }

    protected void showTime() {
        OnSlideBuildAdapter listener = (OnSlideBuildAdapter) timetableView.onSlideBuildListener();
        String[] showTimes = new String[Course.MAX_STEPS];
        for (int i = 0; i < Course.MAX_STEPS; i++) {
            showTimes[i] = Course.START_TIMES[i] + "\n" + Course.END_TIMES[i];
        }
        listener.setTimes(showTimes)
                .setTimeTextColor(Color.BLACK);
        timetableView.updateSlideView();
    }

    @Override
    protected void onResume() {
        super.onResume();

        boolean use_chi_icon = preferences.getBoolean("use_chi_icon", false);
        setIcon(use_chi_icon);

        courseManager.updateStatus();

        showWeek = timetableView.curWeek();
        updateView();
    }

    private void setIcon(boolean use_chi_icon) {
        String disable_activity = use_chi_icon ? ".MainActivity" : ".MainAliasActivity";
        String enable_activity = (!use_chi_icon) ? ".MainActivity" : ".MainAliasActivity";
        PackageManager packageManager = getPackageManager();
        packageManager.setComponentEnabledSetting(new ComponentName(this, getPackageName() +
                disable_activity), PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager
                .DONT_KILL_APP);
        packageManager.setComponentEnabledSetting(new ComponentName(this, getPackageName() +
                enable_activity), PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager
                .DONT_KILL_APP);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_login) {
            startActivity(new Intent(MainActivity.this, LoginActivity.class));
        } else if (id == R.id.action_settings) {
            startActivity(new Intent(MainActivity.this, SettingsActivity.class));
        } else if (id == R.id.action_sync) {
            courseManager.updateCourses(new Handler(Looper.getMainLooper()) {
                @Override
                public void handleMessage(@NonNull Message msg) {
                    super.handleMessage(msg);
                    if (msg.what == CourseManager.SUCCEEDED) {
                        updateView();
                        Toast.makeText(MainActivity.this, R.string.succeed_in_sync, Toast.LENGTH_SHORT).show();
                    } else if (msg.what == CourseManager.UNLOGIN) {
                        Toast.makeText(MainActivity.this, R.string.please_log_in, Toast.LENGTH_SHORT).show();
                    } else if (msg.what == CourseManager.FAILED) {
                        Toast.makeText(MainActivity.this, R.string.failed, Toast.LENGTH_SHORT).show();
                    }
                }
            });

        } else if (id == R.id.action_expand) {
            if (weekView.isShowing())
                item.setIcon(R.mipmap.ic_expand_more_white_24dp);
            else
                item.setIcon(R.mipmap.ic_expand_less_white_24dp);

            weekView.isShow(!weekView.isShowing());
        }

        return true;
    }

    private void updateView() {
        courseManager.loadCourses();
        timetableView.source(courseManager.getCourses()).updateView();
        weekView.source(courseManager.getCourses()).showView();
    }

    private void updateWeek() {
        weekManager.updateCurWeek();
        timetableView.curWeek(weekManager.getCurWeek());
        weekView.curWeek(weekManager.getCurWeek());
    }
}

File: app/src/main/java/com/dujiajun/courseblock/MainActivity.kt
package com.dujiajun.courseblock

import android.content.ComponentName
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.preference.PreferenceManager
import com.dujiajun.courseblock.constant.PreferenceKey
import com.dujiajun.courseblock.helper.CourseManager
import com.dujiajun.courseblock.helper.WeekManager
import com.dujiajun.courseblock.model.Course
import com.zhuangfei.timetable.TimetableView
import com.zhuangfei.timetable.listener.IWeekView.OnWeekItemClickedListener
import com.zhuangfei.timetable.listener.OnSlideBuildAdapter
import com.zhuangfei.timetable.listener.OnSpaceItemClickAdapter
import com.zhuangfei.timetable.view.WeekView

open class MainActivity : AppCompatActivity() {
    private lateinit var preferences: SharedPreferences
    private lateinit var courseManager: CourseManager
    private lateinit var timetableView: TimetableView
    private lateinit var weekView: WeekView
    private lateinit var weekManager: WeekManager
    private var showWeek = 1
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        preferences = PreferenceManager.getDefaultSharedPreferences(this)
        weekManager = WeekManager.getInstance(this)
        courseManager = CourseManager.getInstance(applicationContext)
        timetableView = findViewById(R.id.id_timetableView)
        weekView = findViewById(R.id.id_weekview)
        weekView.curWeek(weekManager.getCurWeek())
            .callback(object : OnWeekItemClickedListener {
                override fun onWeekClicked(week: Int) {
                    val cur = timetableView.curWeek()
                    timetableView.onDateBuildListener().onUpdateDate(cur, week)
                    timetableView.changeWeekOnly(week)
                    showWeek = week
                }
            })
            .hideLeftLayout()
            .itemCount(Course.MAX_WEEKS)
            .isShow(false).showView()
        val showWeekend = preferences.getBoolean(PreferenceKey.SHOW_WEEKEND, true)
        val showNotCurWeek = preferences.getBoolean(PreferenceKey.SHOW_NOT_CUR_WEEK, true)
        val showTime = preferences.getBoolean(PreferenceKey.SHOW_COURSE_TIME, true)
        timetableView.curWeek(weekManager.getCurWeek())
            .isShowNotCurWeek(showNotCurWeek)
            .isShowWeekends(showWeekend)
            .maxSlideItem(Course.MAX_STEPS)
            .callback(object : OnSpaceItemClickAdapter() {
                override fun onSpaceItemClick(day: Int, start: Int) {
                    val intent = Intent(this@MainActivity, CourseActivity::class.java)
                    intent.putExtra("day", day + 1)
                    intent.putExtra("start", start)
                    intent.putExtra("week", showWeek)
                    intent.putExtra("action", CourseActivity.ACTION_INSERT)
                    startActivity(intent)
                }
            })
            .callback { _: View?, day: Int, start: Int ->
                val c = courseManager.findCourseByDayAndStart(showWeek, day, start)
                if (c != null) {
                    val intent = Intent(this@MainActivity, CourseActivity::class.java)
                    intent.putExtra("action", CourseActivity.ACTION_DETAIL)
                    intent.putExtra("course", c.id)
                    startActivity(intent)
                }
            }
        if (showTime) showTime()
        updateView()
        updateWeek()
    }

    private fun showTime() {
        val listener = timetableView.onSlideBuildListener() as OnSlideBuildAdapter
        val showTimes = Array(Course.MAX_STEPS) { "" }
        for (i in 0 until Course.MAX_STEPS) {
            showTimes[i] = """
                ${courseManager.startTimes[i]}
                ${courseManager.endTimes[i]}
                """.trimIndent()
        }
        listener.setTimes(showTimes).setTimeTextColor(Color.BLACK)
        timetableView.updateSlideView()
    }

    override fun onResume() {
        super.onResume()
        val useChiIcon = preferences.getBoolean(PreferenceKey.USE_CHI_ICON, false)
        setIcon(useChiIcon)
        courseManager.updateStatus()
        showWeek = timetableView.curWeek()
        updateView()
        val showTime = preferences.getBoolean(PreferenceKey.SHOW_COURSE_TIME, true)
        if (showTime) showTime()
    }

    private fun setIcon(useChiIcon: Boolean) {
        val disableActivity = if (useChiIcon) ".MainActivity" else ".MainAliasActivity"
        val enableActivity = if (!useChiIcon) ".MainActivity" else ".MainAliasActivity"
        val packageManager = packageManager
        packageManager.setComponentEnabledSetting(
            ComponentName(this, packageName + disableActivity),
            PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP
        )
        packageManager.setComponentEnabledSetting(
            ComponentName(this, packageName + enableActivity),
            PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP
        )
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_login -> {
                startActivity(Intent(this@MainActivity, LoginActivity::class.java))
            }

            R.id.action_settings -> {
                startActivity(Intent(this@MainActivity, SettingsActivity::class.java))
            }

            R.id.action_sync -> {
                courseManager.updateCourses(object : Handler(Looper.getMainLooper()) {
                    override fun handleMessage(msg: Message) {
                        super.handleMessage(msg)
                        when (msg.what) {
                            CourseManager.SUCCEEDED -> {
                                updateView()
                                Toast.makeText(
                                    this@MainActivity,
                                    R.string.succeed_in_sync,
                                    Toast.LENGTH_SHORT
                                ).show()
                            }

                            CourseManager.UNLOGIN -> {
                                Toast.makeText(
                                    this@MainActivity,
                                    R.string.please_log_in,
                                    Toast.LENGTH_SHORT
                                ).show()
                            }

                            CourseManager.FAILED -> {
                                Toast.makeText(
                                    this@MainActivity,
                                    R.string.failed,
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        }
                    }
                })
            }

            R.id.action_expand -> {
                if (weekView.isShowing) item.setIcon(R.mipmap.ic_expand_more_white_24dp) else item.setIcon(
                    R.mipmap.ic_expand_less_white_24dp
                )
                weekView.isShow(!weekView.isShowing)
            }
        }
        return true
    }

    private fun updateView() {
        courseManager.loadCourses()
        timetableView.source(courseManager.courses).updateView()
        weekView.source(courseManager.courses).showView()
    }

    private fun updateWeek() {
        weekManager.updateCurWeek()
        timetableView.curWeek(weekManager.getCurWeek())
        weekView.curWeek(weekManager.getCurWeek())
    }
}